src/rules/objectComparisonRule.ts

```diff
@@ -0,0 +1,278 @@
+/**
+ * @license
+ * Copyright 2019 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { isBinaryExpression, isTypeFlagSet, isUnionType } from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+import { codeExamples } from "./code-examples/objectComparison.examples";
+
+const OPTION_ALLOW_OBJECT_EQUAL_COMPARISON = "allow-object-equal-comparison";
+const OPTION_ALLOW_STRING_ORDER_COMPARISON = "allow-string-order-comparison";
+
+const enum TypeKind {
+    Any = 0,
+    Number = 1,
+    Enum = 2,
+    String = 3,
+    Boolean = 4,
+    NullOrUndefined = 5,
+    Object = 6,
+}
+
+const typeNames = {
+    [TypeKind.Any]: "any",
+    [TypeKind.Number]: "number",
+    [TypeKind.Enum]: "enum",
+    [TypeKind.String]: "string",
+    [TypeKind.Boolean]: "boolean",
+    [TypeKind.NullOrUndefined]: "null | undefined",
```

Comment:
Error messages mentioning `null | undefined` for types that are just `undefined` are a little misleading. `--strictNullCheck` projects make a distinction between the `null` and `undefined` types. `null` and `undefined` should be separate messages & test cases.